src/util/timer.rs

```diff
@@ -0,0 +1,242 @@
+    }
+
+    /// Remove a task from the timer. Returns the `TimeoutTask` if found.
+    ///
+    /// If there are many tasks according with the condition, only the first
+    /// task will be removed from pending queue.
+    pub fn remove_task<F>(&mut self, f: F) -> Option<TimeoutTask<T>>
+    where
+        F: Fn(&TimeoutTask<T>) -> bool,
+    {
+        let mut res = None;
+        let mut vec = replace(&mut self.pending, BinaryHeap::new()).into_vec();
+        if let Some((idx, _)) = vec.iter().enumerate().find(|&(_, task)| f(task)) {
+            res = Some(vec.swap_remove(idx));
+        }
+        self.pending = BinaryHeap::from(vec);
+        res
+    }
+
+    /// Get the next `timeout` from the timer.
+    pub fn next_timeout(&mut self) -> Option<Instant> {
+        self.pending.peek().map(|task| task.next_tick)
+    }
+
+    /// Pop a `TimeoutTask` from the `Timer`, which should be tick before `instant`.
+    /// If there is no tasks should be ticked any more, None will be returned.
+    ///
+    /// The normal use case is keeping `pop_task_before` until get `None` in order
+    /// to retreive all avaliable events.
+    pub fn pop_task_before(&mut self, instant: Instant) -> Option<TimeoutTask<T>> {
+        match self.pending.pop() {
+            Some(timeout_task) => if timeout_task.next_tick > instant {
+                self.pending.push(timeout_task);
+                return None;
+            } else {
+                return Some(timeout_task);
+            },
+            None => None,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::sync::mpsc::{self, Sender};
+    use std::sync::mpsc::RecvTimeoutError;
+    use util::worker::{BatchRunnableWithTimer, Builder as WorkerBuilder, Runnable};
+
+    #[derive(Debug, PartialEq, Eq, Copy, Clone)]
+    enum Task {
+        A,
+        B,
+        C,
+    }
+
+    struct Runner {
+        counter: usize,
+        ch: Sender<&'static str>,
+    }
+
+    impl Runnable<&'static str> for Runner {
+        fn run(&mut self, msg: &'static str) {
+            self.ch.send(msg).unwrap();
+        }
+        fn shutdown(&mut self) {
+            self.ch.send("").unwrap();
+        }
+    }
+
+    impl BatchRunnableWithTimer<&'static str, Task> for Runner {
+        fn on_timeout(&mut self, task: TimeoutTask<Task>, timer: &mut Timer<Task>) {
+            match *task.as_ref() {
+                Task::A => self.ch.send("task a").unwrap(),
+                Task::B => self.ch.send("task b").unwrap(),
+                _ => unreachable!(),
+            };
+            if self.counter < 2 {
+                timer.add_task(task.timeout(), task.into_inner());
+            } else {
+                timer.remove_task(|t| t.as_ref() == &Task::B);
+            }
+            self.counter += 1;
+        }
+    }
+
+    #[test]
+    fn test_timer() {
+        let mut timer = Timer::new(10);
+        timer.add_task(Duration::from_millis(20), Task::A);
```

Comment:
the expired time is too small, I guess if we run on the slow machine or the thread is blocked, we may fail the test.

Should use a larger time.